src/eradius_client.erl

```diff
@@ -125,42 +126,86 @@ send_remote_request(Node, {IP, Port, Secret}, Request, Options) when ?GOOD_CMD(R
             update_client_response(bad_authenticator, MetricsInfo, Request),
             ?LOG(error, "~s INF: Noreply for request ~p. Could not decode the request, reason: ~s", [printable_peer(ServerIP, Port), Request, Reason]),
-            noreply;
+            maybe_failover(Request, noreply, {ServerIP, Port}, Options);
         {bad_pdu, "unknown request type" = Reason} ->
             update_client_response(unknown_req_type, MetricsInfo, Request),
             ?LOG(error, "~s INF: Noreply for request ~p. Could not decode the request, reason: ~s", [printable_peer(ServerIP, Port), Request, Reason]),
             noreply;
         {bad_pdu, Reason} ->
             update_client_response(dropped, MetricsInfo, Request),
             ?LOG(error, "~s INF: Noreply for request ~p. Could not decode the request, reason: ~s", [printable_peer(ServerIP, Port), Request, Reason]),
-            noreply;
+            maybe_failover(Request, noreply, {ServerIP, Port}, Options);
         Decoded ->
             update_client_response(Decoded#radius_request.cmd, MetricsInfo, Request),
             {ok, Response, Authenticator}
     end;
-proceed_response(Request, Response, _Peer, TS1, MetricsInfo) ->
+
+proceed_response(Request, Response, {_ServerName, {ServerIP, Port}}, TS1, MetricsInfo, Options) ->
     update_client_responses(MetricsInfo),
     update_client_request(Request#radius_request.cmd, MetricsInfo, eradius_lib:timestamp(milli_seconds) - TS1, Request),
-    Response.
+    maybe_failover(Request, Response, {ServerIP, Port}, Options).
+
+maybe_failover(Request, Response, {ServerIP, Port}, Options) ->
+    case proplists:get_value(failover, Options, []) of
+        [] ->
+            Response;
+        [{NewServerIP, NewPort, NewSecret, _ServerRetries} | UpstreamServers] ->
+            handle_failed_request(Request, {ServerIP, Port}, {NewServerIP, NewPort, NewSecret}, UpstreamServers, Response, Options);
+        [{NewServerIP, NewPort, NewSecret} | UpstreamServers] ->
+            handle_failed_request(Request, {ServerIP, Port}, {NewServerIP, NewPort, NewSecret}, UpstreamServers, Response, Options)
+    end.
+
+handle_failed_request(Request, {ServerIP, Port}, {NewServerIP, NewPort, NewSecret}, UpstreamServers, Response, Options) ->
+    case ets:lookup(?MODULE, {ServerIP, Port}) of
+        [{{ServerIP, Port}, Retries, InitialRetries}] ->
+            RetriesDone = proplists:get_value(retries, Options, ?DEFAULT_RETRIES),
+            if (Retries - RetriesDone) =< 0 ->
+                    ets:delete(?MODULE, {ServerIP, Port}),
+                    Timeout = application:get_env(eradius, unreachable_timeout, 60),
+                    timer:apply_after(Timeout * 1000, ?MODULE, restore_upstream_server,
+                                      [{ServerIP, Port, InitialRetries, InitialRetries}]);
+               true ->
+                    ets:update_counter(?MODULE, {ServerIP, Port}, -RetriesDone)
```

Comment:
`retries` option could be passed to `eradius_client:send_radius_request/3`, so client will try to send `N` RADIUS requests to the given server where `N` is value of the `retries`. So it could be less than `1`, that is why `-RetriesDone` here instead of `-1`. But again, yes, the name could be not very clear here so I will rename it to `FailedTries` or so and will put the comment.

Reply:
added comment